lib/toaster/toastergui/views.py

```diff
@@ -0,0 +1,1935 @@
+        or_values = kv[1].split(OR_VALUE_SEPARATOR)
+        query = None
+        for key, val in zip(or_keys, or_values):
+            x = __get_q_for_val(key, val)
+            query = query | x if query else x
+
+        and_query = and_query & query if and_query else query
+
+    return and_query
+
+def _get_toggle_order(request, orderkey, toggle_reverse = False):
+    if toggle_reverse:
+        return "%s:+" % orderkey if request.GET.get('orderby', "") == "%s:-" % orderkey else "%s:-" % orderkey
+    else:
+        return "%s:-" % orderkey if request.GET.get('orderby', "") == "%s:+" % orderkey else "%s:+" % orderkey
+
+def _get_toggle_order_icon(request, orderkey):
+    if request.GET.get('orderby', "") == "%s:+"%orderkey:
+        return "down"
+    elif request.GET.get('orderby', "") == "%s:-"%orderkey:
+        return "up"
+    else:
+        return None
+
+# we check that the input comes in a valid form that we can recognize
+def _validate_input(field_input, model):
+
+    invalid = None
+
+    if field_input:
+        field_input_list = field_input.split(FIELD_SEPARATOR)
+
+        # Check we have only one colon
+        if len(field_input_list) != 2:
+            invalid = "We have an invalid number of separators: " + field_input + " -> " + str(field_input_list)
+            return None, invalid
+
+        # Check we have an equal number of terms both sides of the colon
+        if len(field_input_list[0].split(AND_VALUE_SEPARATOR)) != len(field_input_list[1].split(AND_VALUE_SEPARATOR)):
+            invalid = "Not all arg names got values"
+            return None, invalid + str(field_input_list)
+
+        # Check we are looking for a valid field
+        valid_fields = [f.name for f in model._meta.get_fields()]
+        for field in field_input_list[0].split(AND_VALUE_SEPARATOR):
+            if True in [field.startswith(x) for x in valid_fields]:
+                break
+        else:
+           return None, (field, valid_fields)
+
+    return field_input, invalid
+
+# uses search_allowed_fields in orm/models.py to create a search query
+# for these fields with the supplied input text
+def _get_search_results(search_term, queryset, model):
+    search_object = None
+    for st in search_term.split(" "):
+        queries = None
+        for field in model.search_allowed_fields:
+            query = Q(**{field + '__icontains': st})
+            queries = queries | query if queries else query
+
+        search_object = search_object & queries if search_object else queries
+    queryset = queryset.filter(search_object)
+
+    return queryset
+
+
+# function to extract the search/filter/ordering parameters from the request
+# it uses the request and the model to validate input for the filter and orderby values
+def _search_tuple(request, model):
+    ordering_string, invalid = _validate_input(request.GET.get('orderby', ''), model)
+    if invalid:
+        raise BaseException("Invalid ordering model:" + str(model) + str(invalid))
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Using `Exception` and `BaseException` can make your code prone to errors and difficult to maintain. Instead, we recommend using one of the [Built-in Exceptions](https://docs.python.org/3/library/exceptions.html) or creating a custom exception class that is derived from `Exception` or one of its subclasses.



Similar issue at line number 311.